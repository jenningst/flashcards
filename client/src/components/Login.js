import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link, withRouter } from 'react-router-dom';
import { WaveSpinner } from 'react-spinners-kit';
import { Title2, Caption3 } from '../components/elements/Text';
import { Input } from '../components/elements/Input';
import { PrimaryButton } from './elements/Button';
import { loginWithEmail } from '../contexts/auth-context';
import ROUTE_CONFIG from '../constants/route-config';

Login.propTypes = {
  history: PropTypes.object,
};

function Login({ history }) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState(null);
  const [submitting, setSubmitting] = useState(false);
  const handleEmailChange = e => setEmail(e.target.value);
  const handlePasswordOneChange = e => setPassword(e.target.value);

  const handleLogin = async (e) => {
    e.preventDefault();
    // render spinner and attempt user login
    setSubmitting(true);
    try {
      await loginWithEmail(email, password);
      history.push(ROUTE_CONFIG.auth.DASHBOARD);
    } catch (err) {
      switch (err.code) {
        case 'auth/user-not-found':
          setError(`No account found for email ${email}`);
          setPassword('');
          setSubmitting(false);
          break;
        case 'auth/wrong-password':
          setError(`Oops! The password you entered is incorrect.`);
          setPassword('');
          setSubmitting(false);
          break;
        default:
          setError(err.message);
          setPassword('');
          setSubmitting(false);
          break;
      }
    }
  };

  // TODO: provide tooltip for password requirements
  // TODO: probably shouldn't be validating their passwords just yet...
  const invalidInputs = !email || !password;

  return (
    <LoginPageWrapper className="Login">
      <FormWrapper 
        className={`form-wrapper${submitting ? '--submitting' : ''}`}
      >
        <LoginForm className="Login__form">
          <Title2 className="Login__title form-title">Login</Title2>
          <ErrorMessageBox className="Login__errors">
            <Caption3>{error}</Caption3>
          </ErrorMessageBox>
          <Input
            className="Login__input-email"
            aria-label="email"
            id="email"
            onChange={e => handleEmailChange(e)}
            placeholder="Email Address"
            value={email}
          />
          <Input
            className="Login__input-password"
            aria-label="password"
            id="password"
            onChange={e => handlePasswordOneChange(e)}
            placeholder="Password"
            value={password}
          />
          <div className="ancillary-form-controls">
            <div className="Login__toggle-remember-me">
              <Caption3>
                Remember me
              </Caption3>
            </div>
            <Link to={"/forgot-password"}><Caption3>Forgot Password</Caption3></Link>
          </div>
          <LoginButton
            className="Login__btn-submit"
            type="primary"
            ripple
            onClick={e => handleLogin(e)}
            disabled={invalidInputs ? true : false}
          >
            Log In
          </LoginButton>
        </LoginForm>
        <span>
          <Caption3>New to Flashcards? <Link to={"/signup"}>Sign up</Link></Caption3>
        </span>
      </FormWrapper>
      <SpinnerWrapper
        className={`loading-spinner-wrapper${submitting ? '--visible' : 'hidden'}`}
      >
        {submitting
          ? <>
              <WaveSpinner
                className="Login__spinner"
                size={30}
                loading={submitting}
                color={'#939CE8'}
              />
              <Title2>Logging In</Title2>
            </>
          : null
        }
      </SpinnerWrapper>
    </LoginPageWrapper>
  );
};

const LoginPageWrapper = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;
  height: 100%;
`;

const SpinnerWrapper= styled.div`
  position: absolute;
  box-sizing: border-box;
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;

  opacity: 0.9;

  &[class*="hidden"] {
    z-index: -1;
  }

  &[class*="visible"] {
    z-index: 99999;
  }
`;

const FormWrapper = styled.div`
  box-sizing: border-box;
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 2rem;
  background: ${props => props.theme.color.background.pureWhite};

  a {
    color: ${props => props.theme.color.font.primary};
    text-decoration: none;
    outline: none;
    border-bottom: 1px solid transparent;

    &:hover {
      transition: color .2s ease-out
      border-bottom-color .2s ease-out;
      border-bottom: 1px solid ${props => props.theme.color.border.primary};
    }
  }

  @media(min-width: 376px) {
    min-width: 320px;
    max-width: 450px;
    height: auto;
    box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);
    border-radius: 1rem;
  }

  &[class*="submitting"] {
    opacity: 0.2;
  }
`;

const ErrorMessageBox = styled.span`
  width: 100%;
  height: 2rem;
  color: ${props => props.theme.color.font.danger};
`;

const LoginForm = styled.form`
  box-sizing: border-box;
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 100%;

  .form-title {
    width: 100%;
    text-align: left;
    margin-bottom: 1.5rem;
  }

  .ancillary-form-controls {
    display: flex;
    justify-content: space-between;
    width: 100%;
    margin-top: 2rem;
  }
`;

const LoginButton = styled(PrimaryButton)`
  margin: 1.5rem;
  width: 100%;
`;

export default withRouter(Login);