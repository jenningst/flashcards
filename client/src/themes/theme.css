/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/