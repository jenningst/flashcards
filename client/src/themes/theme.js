export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};