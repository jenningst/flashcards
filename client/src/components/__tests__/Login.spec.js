import React from 'react';
import { StaticRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';

import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';

import Login from '../Login';

afterEach(cleanup);

const renderComponent = ({ theme }) =>
  render(
    <ThemeProvider theme={theme}>
      <StaticRouter>
        <Login />
      </StaticRouter>
    </ThemeProvider>
  );

describe('<Login /> spec', () => {
  it('assert component matches snapshot', () => {
    const { asFragment } = renderComponent({ theme: lightTheme });
    expect(asFragment()).toMatchSnapshot();
  });

  it('assert it initially renders an empty input for email', () => {
    const { getByLabelText } = renderComponent({ theme: lightTheme });
    const emailElement = getByLabelText(/^email address$/i);
    expect(emailElement).toBeInTheDocument();
    expect(emailElement).toBeEmpty();
  });

  it('assert it initially renders an empty input for password one', () => {
    const { getByLabelText } = renderComponent({ theme: lightTheme });
    const passwordElement = getByLabelText(/^password$/i);
    expect(passwordElement).toBeInTheDocument();
    expect(passwordElement).toBeEmpty();
  });

  it('assert it initially renders a disabled submit button', () => {
    const { getByText } = renderComponent({ theme: lightTheme });
    expect(getByText(/^log in$/i)).toBeDisabled();
  });

  it('assert it renders an enabled submit button when all inputs have valid values', () => {
    const { getByLabelText, getByText } = renderComponent({ theme: lightTheme });
    const emailElement = getByLabelText(/^email address$/i);
    const passwordOneElement = getByLabelText(/^password$/i);
    const mockEmailEvent = { target: { value: 'test@gmail.com' }};
    const mockPasswordEvent = { target: { value: 'ABC123' }};
    fireEvent.change(emailElement, mockEmailEvent);
    fireEvent.change(passwordOneElement, mockPasswordEvent);
    expect(getByText(/^log in$/i)).toBeEnabled();
  });

  it('assert it renders a link to login page', () => {
    const { getByText } = renderComponent({ theme: lightTheme });
    expect(getByText(/^don't have an account/i)).toBeInTheDocument();
  });

  it('assert link redirects to sign up page', () => {
    const { getByText } = renderComponent({ theme: lightTheme });
    expect(getByText(/^don't have an account/i)).toHaveAttribute('href', '/signup');
  });
});