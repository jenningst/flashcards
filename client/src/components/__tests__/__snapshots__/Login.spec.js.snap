// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> spec assert component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Login sc-cSHVUG cnhTFu"
  >
    <div
      class="form-wrapper sc-chPdSV eTDCRq"
    >
      <form
        class="Login__form sc-kGXeez jlvJeS"
      >
        <h2
          class="Login__title form-title sc-htpNat dYSOWx"
        >
          Login
        </h2>
        <span
          class="Login__errors sc-kgoBCf fIjuNl"
        >
          <h6
            class="sc-iwsKbI iMjCin"
          />
        </span>
        <input
          aria-label="email"
          class="sc-gZMcBi Login__input-email sc-gqjmRU hOlNGk"
          id="email"
          placeholder="Email Address"
          value=""
        />
        <input
          aria-label="password"
          class="sc-gZMcBi Login__input-password sc-gqjmRU hOlNGk"
          id="password"
          placeholder="Password"
          value=""
        />
        <div
          class="ancillary-form-controls"
        >
          <div
            class="Login__toggle-remember-me"
          >
            <h6
              class="sc-iwsKbI iMjCin"
            >
              Remember me
            </h6>
          </div>
          <a
            href="/forgot-password"
          >
            <h6
              class="sc-iwsKbI iMjCin"
            >
              Forgot Password
            </h6>
          </a>
        </div>
        <button
          class="sc-VigVT sc-jTzLTM Login__btn-submit sc-kpOJdX jJDPho"
          disabled=""
          type="primary"
        >
          Log In
        </button>
      </form>
      <span>
        <h6
          class="sc-iwsKbI iMjCin"
        >
          New to Flashcards? 
          <a
            href="/signup"
          >
            Sign up
          </a>
        </h6>
      </span>
    </div>
    <div
      class="loading-spinner-wrapperhidden sc-kAzzGY atzHR"
    />
  </div>
</DocumentFragment>
`;
